### PR TITLE
Optimize is_leader checks by caching temporarily

### PR DESCRIPTION
**Pull request by dshcherb**
In order to avoid is-leader hook tool executions on every write to
application relation data bags caching of is-leader return values is
introduced. It relies on the implementation detail in Juju where it
grants a lease for 30 seconds when either of the following happens:

1) a unit agent renews the lease;
2) is-leader is called;
3) leader-set is called.

### CODE REVIEW

File: op/model.py [hunk 1]
```diff
@@ -313,4 +323,14 @@ def config_get(self):
         return self._run('config-get')
 
     def is_leader(self):
-        return self._run('is-leader')
+        """Obtain the current leadership status for the unit the charm code is executing on.
+
+        The value is cached for the duration of a lease which is 30s in Juju.
+        """
+        if datetime.timedelta(seconds=time.monotonic() - self._leader_check_time) > self.LEASE_REFRESH_PERIOD or self._is_leader is None:
```
Comment (jameinel) [inline op/model.py]: as a minor optimization, and as a safety measure, you probably want to do:

now = time.monotonic()
time_since_check = datetime.timedelta(second=now - self._leader_check_time)
if (self._is_leader is None or time_since_check > self.LEASE_REFRESH_PERIOD):
    self._is_leader = self._run('is-leader')
    self._leader_check_time = now
return self._is_leader

That avoids calling time.monotonic() twice and IMO is a bit easier to read. The comment still holds, since we call time.monotonic() before the IsLeader call.

File: test/test_model.py [hunk 2]
```diff
@@ -212,6 +213,26 @@ def check_remote_units():
...
+    def test_is_leader_refresh(self):
+        self.backend = op.model.ModelBackend()
+        self.model = op.model.Model('myapp/0', ['db0', 'db1', 'db2'], self.backend)
+
+        # A sanity check.
+        self.assertGreater(time.monotonic(), op.model.ModelBackend.LEASE_REFRESH_PERIOD.total_seconds())
+
+        fake_script(self, 'is-leader', 'echo true')
+        self.assertTrue(self.model.unit.is_leader())
+
```
Comment (jameinel) [inline test/test_model.py]: Don't you want a
```
# Change the return value, but see that we use the cached value
# In practice, we know the leadership won't change before we
# would timeout and check again.
fake_script(self, 'is-leader', 'echo false')
self.assertTrue(self.model.unit.is_leader())
```
Otherwise, while you check that setting _leader_check_time = 0 *does* trigger a refresh, nothing in this test would fail if we just always refreshed. (eg, take out the caching, and I believe this test still passes)

File: op/model.py [hunk 3]
```diff
@@ -259,10 +263,16 @@ class RelationNotFound(ModelError):
 
 
 class ModelBackend:
+
+    LEASE_REFRESH_PERIOD = datetime.timedelta(seconds=30)
```
Comment (jameinel) [inline op/model.py]: Is LEASE_CACHE_TIME easier to understand? (LEASE_CACHE_EXPIRY_TIME) it does start to get a bit wordy.
Not clearly better IMOl, just a brainstorm of words we could use.

File: op/model.py (hunk 1 above)
Comment (dshcherb) [reply op/model.py]: Agreed, thanks.

File: test/test_model.py (hunk 2 above)
Comment (dshcherb) [reply test/test_model.py]: Good call.

File: op/model.py (hunk 3 above)
Comment (dshcherb) [reply op/model.py]: How about LEASE_RENEWAL_PERIOD?

I want to be explicit about it being a period of time, rather than a _point in time_. Usage of timedelta suggests that it's the case but the type is not exposed in the variable name.

**Comment by dshcherb** [comment]
Rebased on top of the latest master (had to update the first commit to avoid test failures).
